File: frontend/lib/h3-audio-guidance.ts
import type { H3Scene } from './h3-projects'

export function composeH3AudioPrompt(scene: Pick<H3Scene, 'prompt' | 'audio_mode' | 'no_speech' | 'no_music' | 'custom_audio_instruction'>): string {
  const prompt = scene.prompt.trim()
  if (!prompt) return ''
  if (scene.audio_mode === 'silent') return `${prompt} No speech, no dialogue, no voices, no vocalizations, no music, no soundtrack, no background score, no singing, no musical elements, no ambient sound.`
  const parts = [prompt]
  if (scene.audio_mode === 'natural_ambience') parts.push('Natural environmental ambience appropriate to the scene.')
  if (scene.custom_audio_instruction.trim()) parts.push(scene.custom_audio_instruction.trim())
  if (scene.no_speech) parts.push('No speech, no dialogue, no voices, no vocalizations.')
  if (scene.no_music) parts.push('No music, no soundtrack, no background score, no singing, no musical elements.')
  return parts.join(' ')
}

export function h3AudioSummary(scene: Pick<H3Scene, 'audio_mode' | 'no_speech' | 'no_music'>): string {
  if (scene.audio_mode === 'silent') return 'Silent'
  const labels = [scene.audio_mode === 'natural_ambience' ? 'Natural ambience' : 'Dialogue']
  if (scene.no_speech) labels.push('No speech')
  if (scene.no_music) labels.push('No music')
  return labels.join(' · ')
}

File: backend/services/h3_audio_guidance.py
"""Deterministic, local-only audio prompt guidance for MiniMax H3 scenes."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

AudioMode = Literal["natural_ambience", "dialogue", "silent"]


@dataclass(frozen=True)
class H3AudioGuidance:
    mode: AudioMode = "natural_ambience"
    no_speech: bool = False
    no_music: bool = False
    custom_instruction: str = ""


def compose_h3_prompt(original_prompt: str, guidance: H3AudioGuidance) -> str:
    """Append predictable audio instructions once, without changing user prose."""
    original = original_prompt.strip()
    if not original:
        raise ValueError("A scene prompt is required.")

    parts = [original]

    def append_once(instruction: str) -> None:
        if instruction.lower() not in original.lower() and all(instruction.lower() not in item.lower() for item in parts):
            parts.append(instruction)
    custom = guidance.custom_instruction.strip()
    if guidance.mode == "silent":
        # Silence has deterministic precedence over every audio option and custom sound.
        append_once("No speech, no dialogue, no voices, no vocalizations, no music, no soundtrack, no background score, no singing, no musical elements, no ambient sound.")
    else:
        if guidance.mode == "natural_ambience":
            append_once("Natural environmental ambience appropriate to the scene.")
        # Dialogue intentionally adds no invented words and never suppresses user dialogue.
        if custom:
            append_once(custom)
        if guidance.no_speech:
            append_once("No speech, no dialogue, no voices, no vocalizations.")
        if guidance.no_music:
            append_once("No music, no soundtrack, no background score, no singing, no musical elements.")
    return " ".join(parts)


def audio_summary(guidance: H3AudioGuidance) -> str:
    if guidance.mode == "silent":
        return "Silent"
    labels = ["Natural ambience" if guidance.mode == "natural_ambience" else "Dialogue"]
    if guidance.no_speech:
        labels.append("No speech")
    if guidance.no_music:
        labels.append("No music")
    return " · ".join(labels)
